File: docs/algorithms/code/algorithms/removeNthFromEnd.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  if (!head) {
    return null;
  }

  let temp = head;

  let count = 0;
  while (temp) {
    count++;
    temp = temp.next;
  }

  const targetIndex = count - n;

  if (targetIndex === 0) {
    return head.next;
  }
  temp = head;

  let index = 0;
  while (temp) {
    if (index + 1 === targetIndex) {
      temp.next = temp.next.next;
      break;
    }
    temp = temp.next;
    index++;
  }

  return head;
}

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  const node = new ListNode(0, head);
  let left: ListNode | null = head,
    right: ListNode | null = node;

  for (let i = 0; i < n; i++) {
    left = left.next;
  }

  while (left) {
    left = left.next;
    right = right.next;
  }

  right.next = right.next.next;

  return node.next;
}

File: docs/algorithms/code/algorithms/removeNthFromEnd.go
// 思路： 快慢指针，快指针先走 n 步，然后快慢指针一起走，直到快指针走到头，然后慢指针的下一个节点就是要删除的节点
// 特殊处理：所以我们把慢指针指向一个虚拟头节点，这样就可以处理删除头节点的情况
// 复杂度分析：时间复杂度 O(n)，空间复杂度 O(1)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func removeNthFromEnd(head *ListNode, n int) *ListNode {
  if head == nil {
		return head
	}  

	fast := head;
	tempNode := ListNode{ Val: 0, Next: head }
	slow := &tempNode;

	for i:=0; i<n; i++ {
		fast = fast.Next;
	}

	if fast == nil {
		return head.Next;
	}

	for fast != nil {
		fast = fast.Next;
		slow = slow.Next;
	}

	slow.Next = slow.Next.Next;

	return tempNode.Next;
}